extras/freenode_taxonomy_reader.rb

```diff
@@ -0,0 +1,49 @@
+require 'socket'
+require 'securerandom'
+require 'timeout'
+
+class FreenodeTaxonomyReader
+  NICKSERV = 'NickServ@services'
+  END_MESSGE_REGEX = /End of (.+) taxonomy|(.+) is not registered/
+
+  def initialize(
+        socket_provider: ->() { TCPSocket.new('chat.freenode.net', 6667) },
+        username_provider: ->() { SecureRandom.alphanumeric(16) }
+      )
+    @socket_provider = socket_provider
+    @nick = username_provider.call
+  end
+
+  def for_user(username)
+    _head, *taxonomy_lines = taxonomy_for(username)
+
+    taxonomy_lines.map do |line|
+      line
+        .split(':')[2,3]
+        .map { |t| t.strip }
+    end.to_h
+  end
+
+  private
+
+  def taxonomy_for(username)
```

Comment:
For the record, I dislike how this function fundamentally works; it doesn't understand the protocol, instead basically scraping the lines it receives for the data it wants. This makes it somewhat fragile.

Additionally, there is no error handling; if things break (e.g. it's hitting some connection rate limit, services are temporarily down for maintenance, etc), it doesn't seem to distinguish that from simply getting an empty taxonomy output.

Reply:
Howdy, can you point to a doc about the protocol and format for these messages? I don't know if we want to actually break out a proper parser but I'd like to catch error messages and such if possible.

Reply:
You get a 32-wide, space padded key field, followed by a `: ` and the value field, with a header and footer NOTICE that can be used to track the state of your query.

`:` is illegal in key names, but can be included in values.

```
<< PRIVMSG nickserv :taxonomy kline
>> :NickServ!NickServ@services. NOTICE kline :Taxonomy for kline:
>> :NickServ!NickServ@services. NOTICE kline :abc                             : def
>> :NickServ!NickServ@services. NOTICE kline :foo                             : bar:baz
>> :NickServ!NickServ@services. NOTICE kline :End of kline taxonomy.
```

Reply:
For the protocol in general, RFC1459 has [the detailed syntax](https://tools.ietf.org/html/rfc1459#section-2.3.1). You might get away with some clever use of a string split function for this; it's just the last parameter which makes it a bit more annoying (since it has to allow spaces).

If there's an issue that gets your entire connection rejected, you should receive an `ERROR` message with a parameter that tells you the issue (which might not be useful to display to users, but might be worth logging or something to be able to figure things out if they go wrong).

You might also want to listen to any "numeric", i.e. messages with a three-digit command, in the 400-599 range; those tend to be (usually non-fatal to your connection) errors, such as "no such nick" errors that would imply NickServ is down for maintenance. (IRC being a mess, there are error numerics outside that range, though none I see that'd apply in your use case – with our server software, anyway.)

For the actual taxonomy output, @AbstractBeliefs has you covered. [Note you may also receive an error – as a `NOTICE` – if the user isn't registered](https://github.com/atheme/atheme/blob/f8cd25ff8e35d8fde3d136e1fb5421d0068aa465/modules/nickserv/taxonomy.c#L51).

I really need to make a decent API for this sort of thing at some point …